lib/preface_make/apply.ml

```diff
@@ -0,0 +1,185 @@
+module Core_via_map_and_product (Req : Preface_specs.Apply.WITH_MAP_AND_PRODUCT) =
+struct
+  include Req
+
+  let apply f a = map (fun (f, a) -> f a) @@ product f a
+  let lift2 f x y = apply (map f x) y
+end
+
+module Core_over_functor_via_apply
+    (Req : Preface_specs.Apply.WITH_APPLY)
+    (Functor : Preface_specs.Functor.WITH_MAP with type 'a t = 'a Req.t) =
+struct
+  include Req
+  include Functor
+
+  let product a b = apply (map (fun a b -> (a, b)) a) b
+  let lift2 f x y = apply (map f x) y
+end
+
+module Core_over_functor_via_lift2
+    (Req : Preface_specs.Apply.WITH_LIFT2)
+    (Functor : Preface_specs.Functor.WITH_MAP with type 'a t = 'a Req.t) =
+struct
+  include Req
+  include Functor
+
+  let apply f a = lift2 (fun x -> x) f a
+  let product a b = apply (map (fun a b -> (a, b)) a) b
+end
+
+module Operation (Core : Preface_specs.Apply.CORE) = struct
+  include Functor.Operation (Core)
+
+  let lift = Core.map
+  let lift3 f a b = Core.(apply @@ apply (Core.map f a) b)
+end
+
+module Syntax (Core : Preface_specs.Apply.CORE) = struct
+  type 'a t = 'a Core.t
+
+  let ( let+ ) x f = Core.map f x
+  let ( and+ ) = Core.product
+end
+
+module Infix
+    (Core : Preface_specs.Apply.CORE)
+    (Operation : Preface_specs.Apply.OPERATION with type 'a t = 'a Core.t) =
+struct
+  include Functor.Infix (Core) (Operation)
+
+  let ( <*> ) = Core.apply
+  let ( <**> ) a b = Core.lift2 (fun x f -> f x) a b
+  let ( *> ) a b = Core.lift2 (flip const) a b
+  let ( <* ) a b = Core.lift2 const a b
+end
+
+module Via
+    (Core : Preface_specs.Apply.CORE)
+    (Operation : Preface_specs.Apply.OPERATION)
+    (Infix : Preface_specs.Apply.INFIX)
+    (Syntax : Preface_specs.Apply.SYNTAX) =
+struct
+  include Core
+  include Operation
+  include Syntax
+  include Infix
+  module Infix = Infix
+  module Syntax = Syntax
+end
+
+module Via_map_and_product (Req : Preface_specs.Apply.WITH_MAP_AND_PRODUCT) =
+struct
+  module Core = Core_via_map_and_product (Req)
+  module Operation = Operation (Core)
+  module Syntax = Syntax (Core)
+  module Infix = Infix (Core) (Operation)
+  include Core
+  include Operation
+  include Syntax
+  include Infix
+end
+
+module Via_map_and_apply
+    (Req : Preface_specs.Apply.WITH_APPLY)
+    (Functor : Preface_specs.Functor.WITH_MAP with type 'a t = 'a Req.t) =
+struct
+  module Core = Core_over_functor_via_apply (Req) (Functor)
+  module Operation = Operation (Core)
+  module Syntax = Syntax (Core)
+  module Infix = Infix (Core) (Operation)
+  include Core
+  include Operation
+  include Syntax
+  include Infix
+end
+
+module Via_map_and_lift2
+    (Req : Preface_specs.Apply.WITH_LIFT2)
+    (Functor : Preface_specs.Functor.WITH_MAP with type 'a t = 'a Req.t) =
+struct
+  module Core = Core_over_functor_via_lift2 (Req) (Functor)
+  module Operation = Operation (Core)
+  module Syntax = Syntax (Core)
+  module Infix = Infix (Core) (Operation)
+  include Core
+  include Operation
+  include Syntax
+  include Infix
+end
```

Comment:
Um, the API is not consistent with `Via_map_and_product`. I suggest to have several modules: 

- `Over_functor_via_product`
- `Via_map_and_product`
- `Over_functor_via_apply`
- `Via_map_and_apply`
- `Over_functor_via_lift2`
- `Via_map_and_lift2`

(And the same for the versions of `Core`).

Note that I wonder if it wouldn't be better to do something like this: 

```ocaml
module Over_functor (F: Preface_specs.Functor.CORE) : sig 
   module Via_apply (...)
   (* etc *)
end
```
But that would mean a "big change in the whole Preface library...

Reply:
Done